atc/db/notifications_bus.go

```diff
@@ -62,65 +66,35 @@ func (bus *notificationsBus) Listen(channel string) (chan bool, error) {
 
 	sinks[notify] = struct{}{}
 
-	bus.notificationsL.Unlock()
-
 	return notify, nil
 }
 
-func (bus *notificationsBus) Notify(channel string) error {
-	_, err := bus.conn.Exec("NOTIFY " + channel)
-	return err
-}
-
 func (bus *notificationsBus) Unlisten(channel string, notify chan bool) error {
 	bus.notificationsL.Lock()
+	defer bus.notificationsL.Unlock()
+
 	delete(bus.notifications[channel], notify)
-	lastSink := len(bus.notifications[channel]) == 0
-	bus.notificationsL.Unlock()
 
-	if lastSink {
+	if len(bus.notifications[channel]) == 0 {
 		return bus.listener.Unlisten(channel)
 	}
 
 	return nil
 }
 
-func (bus *notificationsBus) dispatchNotifications() {
+func (bus *notificationsBus) wait() {
 	for {
+
 		notification, ok := <-bus.listener.Notify
 		if !ok {
 			break
 		}
 
-		gotNotification := notification != nil
-
 		bus.notificationsL.Lock()
 
-		if gotNotification {
-			// alert any relevant listeners of notification being received
-			// (nonblocking)
+		if notification != nil {
 			for sink := range bus.notifications[notification.Channel] {
-				select {
-				case sink <- true:
-					// notified of message being received (or queued up)
-				default:
-					// already had notification queued up; no need to handle it twice
-				}
-			}
-		} else {
-			// alert all listeners of connection break so they can check for things
-			// they may have missed
-			for _, sinks := range bus.notifications {
-				for sink := range sinks {
-					select {
-					case sink <- false:
```

Comment:
:thinking: I don't see why this should be removed. This is to cover a possibility where ATC1 is tracking the build, loses connectivity, and a user aborts the build (possibly through another ATC whose connectivity is fine) and ATC1 misses the notification while it reconnects. If we ignore it and don't re-check the condition that ATC1 is looking for, the build won't be aborted.

So removing this means condition-notifiers can miss their conditions, and at least for build aborting that can be pretty important. It lowers resiliency. It's an edge case, but it'll happen eventually.

It's a little hairy but the comments should have made it pretty clear (I'm also kind of sad to see comments go away in these trickier places). Should we bring it back? Or is it actually not needed anymore because of something I haven't seen?

Reply:
It's back now :)